### ISSUE TITLE
check returns 0 while kubectl fails

### ISSUE THREAD

**Issue by x3nb63** [jgriff/k8s-resource#12]
as I know now, there is a communication problem with the K8s API. Finally, I found it using the debug shell cause no error message nor any output message is shown on the concourse WebUI, here:

```
$ ./run.sh shell

[... some lines, building the image ...]

root@0c0698f226a2:/code# cat test/fixtures/my-source.json | bash -x assets/check 
+ set -e

[... many lines ...]

+ echo -e '\n-->' querying k8s cluster ''\''https://kmaster.my.domain:8443'\''' in namespace ''\''cd-stage'\''' for ''\''pods'\''' resources...
+ '[' '' = true ']'
++ jq '[.items[]]'
++ kubectl --server=https://kmaster.my.domain:8443 --token=<token string> --certificate-authority=/tmp/resource-ca_file.Yv0IdF get pods -n cd-stage '--sort-by={.metadata.resourceVersion}' -o json
Unable to connect to the server: dial tcp 172.16.0.111:8443: i/o timeout
+ new_versions=
+ log ''

[... many lines ...]

+ log '\n--> all current version(s) to be emitted are: '
+ test 1 -gt 0
+ case "$1" in
+ break
+ '[' '' = true ']'
+ echo -e '\n-->' all current 'version(s)' to be emitted are:
+ '[' '' = true ']'
+ '[' -z '' ']'
+ log '\n--> producing initial version...'
+ test 1 -gt 0
+ case "$1" in
+ break
+ '[' '' = true ']'
+ echo -e '\n-->' producing initial version...
+ '[' '' = true ']'
++ echo
++ tr '\r\n' ' '
++ jq '. | select(length > 0) | .[0]'
+ initial_version=
+ out='[]'
+ log '\n--> OUT\n[]'
+ test 1 -gt 0
+ case "$1" in
+ break
+ '[' '' = true ']'
+ echo -e '\n-->' 'OUT\n[]'
+ '[' '' = true ']'
+ jq flatten
+ jq -n '[]'
[]
root@0c0698f226a2:/code# echo $?
0
root@0c0698f226a2:/code#
```

I suppose the problem is the exit 0.  Cause WebUI says _checked successfully_ without showing any version string cause an empty array `[]` is returned. As a consequence my jobs remains _Pending_ cause it does not get a version ...

AFAIK the i/o timeout reported by kubectl should fail the check.

(also I dont understand where all the `log "..."` strings go? seeing them on the WebUI would have given me a better idea ...)

**Comment by jgriff** [comment]
Thanks for the issue!  Taking a quick look, it appears the problem is the `kubectl` command is [executed in a subshell](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/check#L31), but the calling shell does not check the return code for a potential non-zero result.

💡 We should put a quick assertion in there on the result of the subshell before continuing.  That would at least allow the `check` to fail, if `kubectl` fails.

As for:
> (also I dont understand where all the log "..." strings go? seeing them on the WebUI would have given me a better idea ...)

All log content is written to the container's `/tmp/resource.log` file (see [here](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/common#L34)), as is standard practice for a Concourse resource.  If you want to interrogate it, you can `fly hijack` the container (be sure to include `--check` if hijacking a `check` container).  

For content we want to also appear in the Concourse web ui, those log statements include a `-p` or `--public` option which will also echo them to `stdout` (what Concourse displays in the web ui).  Examples of this are [`in#37`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/in#L37), [`in#42`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/in#L42), [`in#63`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/in#L63), [`out#23`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/out#L23), and [`out#26`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/out#L26).

However, your particular problem is with the `check` container.  **Note that the `check` container's `stdout` is [treated special](https://concourse-ci.org/implementing-resource-types.html#resource-check), and Concourse expects you to emit only the new resource version JSON to `stdout` from `check`.**  That is why you won't see any logging information from `check` hitting the Concourse ui (and why everything from `check` goes to the log file).  This does make debugging issues with `check` slightly more difficult.  But, that's how the resource contract works.


See [here for how `log` is handled](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/common#L12).

**Comment by x3nb63** [comment]
yes, I too guess its the subshell, as the main shells `set -e` will not notice these, so there is no hard stop. And a subsequent `log` call (or whatever) will very likely produce a new zero return.

thanks for explaining the logging - I was not aware of using hijacking to look into resource containers.

**Comment by jgriff** [comment]
This has been addressed in #17 , and released in [`0.0.6`](https://github.com/jgriff/k8s-resource/releases/tag/v0.0.6).
